fix(tools): correct handling multi-tool in queue

bassamsdata opened the pull request:
## Description

This fixes an issue with multiple tools being called one after another. If one tool gets canceled or runs async, it could end up canceling the remaining tool calls in the queue. Since `error()` already calls `setup()`, which closes the failed tool and starts the next one, nothing else needs to be done.

## Related Issue(s)

**Reasoning:** There’s no need to call `close()` at this point, as it would end up closing the newly started tool rather than the one that encountered the error. The errored tool has already been properly closed during the setup invoked by `error()`.

to reproduce: we can share all the tools with LLM and ask it to test them all at once. 

## Screenshots

<!-- Add screenshots of the changes if applicable, to help visualize the change. -->

## Checklist

- [x] I've read the [contributing](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines and have adhered to them in this PR
- [ ] I've added [test](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md#testing) coverage for this fix/feature
- [x] I've run `make all` to ensure docs are generated, tests pass and my formatting is applied
- [ ] _(optional)_ I've updated `CodeCompanion.has` in the [init.lua](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/init.lua#L239) file for my new feature
- [ ] _(optional)_ I've updated the README and/or relevant docs pages

bassamsdata commented:
It seems that the issue is more complex than I initially thought. I'm encountering problems with user confirmation and the approval/rejection process when multiple tools are in the queue. For example, the 'insert_edit_into_file' action is being rejected, but it still appears as if it's been accepted in the chat buffer.

cairijun commented:
It seems #1852 has never thoroughly resolved #1850 . Simply change the `read_file` to async can partially reproduce the issue:

```patch
diff --git a/lua/codecompanion/interactions/chat/tools/builtin/read_file.lua b/lua/codecompanion/interactions/chat/tools/builtin/read_file.lua
index e444d8390..57273a974 100644
--- a/lua/codecompanion/interactions/chat/tools/builtin/read_file.lua
+++ b/lua/codecompanion/interactions/chat/tools/builtin/read_file.lua
@@ -125,8 +125,10 @@ return {
     ---@param args table The arguments from the LLM's tool call
     ---@param input? any The output from the previous function call
     ---@return { status: "success"|"error", data: string }
-    function(self, args, input)
-      return read(args)
+    function(self, args, input, output_handler)
+      vim.schedule(function()
+        output_handler(read(args))
+      end)
     end,
   },
   schema = {
```

With async cmds, all 3 tools are executed, but only the first call's output handler is called:
<img width="450" height="447" alt="1" src="https://github.com/user-attachments/assets/494e769e-2e73-4fbc-9085-50043b5f9dc9" />

With sync cmds, all three errors are reported:
<img width="460" height="432" alt="image" src="https://github.com/user-attachments/assets/f1926b38-e6c7-4671-afc3-b8ba9fa843e7" />

bassamsdata commented:
@olimorris, to make these errors easier to reproduce, I found a straightforward approach using Copilot (GPT-4.1). Simply navigate to the `codecompanion` directory and use the following prompt (please adjust the file path as needed).

```markdown
hello

@{files}
@{cmd_runner}

please test all tools all together since I'm doing a testing for those tools.
for insert_edit_into_file you can add small comment to this file
~/repos/codecompanion.nvim/lua/codecompanion/providers/diff/inline.lua
set oldTEXT to "^" so no need to know file conents.
for some tools i need them in this order:
insert_edit_into_file -> cmd_runner -> create_file -> delete_file
but test them all, all togther in one go please.
NO NEED FOR PLAN. JUST DO IT
```

The LLM will generate a plan regardless; however, after responds with “yes,” several issues begin to surface.

1. While waiting for the diff from `insert_edit_into_file`, all subsequent tools in the queue are canceled (appearing as if the user canceled the tool calls). This seems to occur because the async tool sends a `nil` response to the orchestrator while it is still waiting.
2. There are also multiple issues related to the confirmation prompt that become apparent during this process.
3. Additionally, when attempting to reject certain async tools such as `cmd_runner` or `insert_edit_into_file`, the tool is rejected correctly even though it appears as a successful execution in the chat buffer.

bassamsdata commented:
One more thing I wanted to mention, I’m not entirely sure about this part, so I might be missing something. My understanding is that if a tool is rejected, the other tools in the queue should still be allowed to run. Is that correct?

```lua
-- NOTE: Cancel current tool, then cancel all queued tools
self.output.cancelled(cmd)
self:close()
self:cancel_pending_tools()
```

I’ve made a few new changes in commit [859f9a6](https://github.com/olimorris/codecompanion.nvim/pull/2545/commits/859f9a6059c938a0a6645d1f5d8a35cab1509d3d) and added some comments to make the behavior clearer to my intention.

Thank you

bassamsdata commented:
The current state of this PR addresses all of these issues, although I’m not yet certain that the solutions are the optimal ones.

edit: this PR is just me trying to fix the issues. it's draft and for tacking only.
